flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/ProcTimeInnerJoin.scala

```diff
@@ -0,0 +1,323 @@
+    *
+    * @param timestamp The timestamp of the firing timer.
+    * @param ctx       The ctx to register timer or get current time
+    * @param out       The collector for returning result values.
+    */
+  override def onTimer(
+    timestamp: Long,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+    out: Collector[CRow]): Unit = {
+
+    if (timerState1.value == timestamp) {
+      expireOutTimeRow(
+        timestamp,
+        leftStreamWindowSize,
+        row1MapState,
+        timerState1,
+        ctx
+      )
+    }
+
+    if (timerState2.value == timestamp) {
+      expireOutTimeRow(
+        timestamp,
+        rightStreamWindowSize,
+        row2MapState,
+        timerState2,
+        ctx
+      )
+    }
+  }
+
+  /**
+    * Puts an element from the input stream into state and search the other state to
+    * output records meet the condition, and registers a timer for the current record
+    * if there is no timer at present.
+    */
+  private def processElement(
+    valueC: CRow,
+    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+    out: Collector[CRow],
+    winSize: Long,
+    timerState: ValueState[Long],
+    rowMapState: MapState[Long, JList[Row]],
+    oppoRowMapState: MapState[Long, JList[Row]],
+    oppoWinSize: Long,
+    isPositive: Boolean): Unit = {
+
+    val value = valueC.row
+
+    val curProcessTime = ctx.timerService.currentProcessingTime
+    val oppoExpiredTime =
+      if (oppoWinSize == 0) Long.MinValue else curProcessTime - oppoWinSize
+
+    // only when windowsize != 0, we need to store the element
+    if (winSize != 0) {
+      // register a timer to expire the element
+      if (timerState.value == 0 && winSize != -1) {
+        ctx.timerService.registerProcessingTimeTimer(curProcessTime + winSize + 1)
+        timerState.update(curProcessTime + winSize + 1)
+      }
+
+      var rowList = rowMapState.get(curProcessTime)
+      if (rowList == null) {
+        rowList = new util.ArrayList[Row]()
+      }
+      rowList.add(value)
+      rowMapState.put(curProcessTime, rowList)
+
+    }
+
+    // loop the rightstream elments
+    val oppositeKeyIter = oppoRowMapState.keys().iterator()
+    while (oppositeKeyIter.hasNext) {
+      val eleTime = oppositeKeyIter.next()
+      if (eleTime < oppoExpiredTime) {
+        listToRemove.add(eleTime)
+      } else {
+        val oppoRowList = oppoRowMapState.get(eleTime)
+        var i = 0
+        while (i < oppoRowList.size) {
+          if (isPositive) {
+            compositeOutput(value, oppoRowList.get(i))
+          } else {
+            compositeOutput(oppoRowList.get(i), value)
+          }
+
+          if (filterFunc.filter(outputC.row)) {
+            out.collect(outputC)
+          }
+          i += 1
+        }
+      }
+    }
+
+    // expire records out-of-time
+    var i = listToRemove.size() - 1
+    while (i >= 0) {
+      oppoRowMapState.remove(listToRemove.get(i))
+      listToRemove.remove(i)
```

Comment:
`remove` on an `ArrayList` is inefficient, because it has to shift (i.e., copy) all elements.
We should rather go over the list and call `clear()` when we are done.